ragstack/colbertbase/colbertbase/astra_colbert_embedding.py

```diff
@@ -0,0 +1,224 @@
+    __lr: float = 0.000003,
+    __maxsteps: int = 500000,
+    __nway: int = 2,
+    __use_ib_negatives: bool = False,
+    __reranker: bool = False,
+    __is_cuda: bool = False
+
+    @classmethod
+    def validate_environment(self, values: Dict) -> Dict:
+        """Validate colbert and its dependency is installed."""
+        try:
+            import torch
+            if torch.cuda.is_available():
+                self.__is_cuda = True
+
+        except ImportError as exc:
+            raise ImportError(
+                "Could not import torch library. "
+                "Please install it with `pip install torch`"
+            ) from exc
+
+        return values
+
+    def __init__(
+            self,
+            checkpoint: str = "colbert-ir/colbertv2.0",
+            doc_maxlen: int = 220,
+            nbits: int = 1,
+            kmeans_niters: int = 4,
+            nranks: int = -1,
+            query_maxlen: int = 32,
+            **data: Any,
+    ):
+        self.__cuda = torch.cuda.is_available()
+        total_visible_gpus = 0
+        if self.__cuda:
+            self.__cuda_device_count = torch.cuda.device_count()
+            self.__cuda_device_name = torch.cuda.get_device_name()
+            print(f"nrank {nranks}")
+            if nranks < 1:
+                nranks = self.__cuda_device_count
+            if nranks > 1:
+                total_visible_gpus = self.__cuda_device_count
+            print(f"run on {self.__cuda_device_count} gpus and visible {total_visible_gpus} gpus embeddings on {nranks} gpus")
+        else:
+            if nranks < 1:
+                nranks = 1
+
+        with Run().context(RunConfig(nranks=nranks)):
+            if self.__cuda:
+                torch.cuda.empty_cache()
+            self.colbert_config = ColBERTConfig(
+                doc_maxlen=doc_maxlen,
+                nbits=nbits,
+                kmeans_niters=kmeans_niters,
+                nranks=nranks,
+                checkpoint=checkpoint,
+                query_maxlen=query_maxlen,
+                gpus=total_visible_gpus,
+            )
+        self.__doc_maxlen = doc_maxlen
+        self.__nbits = nbits
+        self.__kmeans_niters = kmeans_niters
+        self.__nranks = nranks
+        print("creating checkpoint")
+        self.checkpoint = Checkpoint(self.colbert_config.checkpoint, colbert_config=self.colbert_config)
+        self.encoder = CollectionEncoder(config=self.colbert_config, checkpoint=self.checkpoint)
+        self.__cuda = torch.cuda.is_available()
+        if self.__cuda:
+            self.checkpoint = self.checkpoint.cuda()
+
+        self.print_memory_stats("ColbertTokenEmbeddings")
+
+    def embed_documents(self, texts: List[str], title: str="") -> List[PassageEmbeddings]:
+        if title == "":
+            title = str(uuid.uuid4())
+        """Embed search docs."""
+        return self.encode(texts, title)
+
+    def embed_query(self, text: str, title: str) -> PassageEmbeddings:
```

Comment:
i'm confused why a query would have a `title` param.

Reply:
Removed. Query does not need title. 
Actually, this piece of code was not used for query. But I made a unit test for this. Since this query embedding does not have padding, I have not tested it.